### ISSUE TITLE
glibc error, invalid pointer when using guards

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11676]
Migrated from [rt.perl.org#100340](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=100340) (status was 'resolved')

Searchable as RT100340$

**Comment by p5pRT** [comment]
### From ostmann@websuche.de

#### Created by ostmann@websuche.de



Attached pure\-perl script produce on all tested versions following glibc error&#8203;:

\*\*\* glibc detected \*\*\* /home/websuche/perl5lib/bin/perl&#8203;: munmap\_chunk\(\)&#8203;: invalid pointer&#8203;: 0x0000000000f72c48 \*\*\*


TESTED VERSIONS&#8203;:
5\.8\.8
5\.10\.0
5\.10\.1
5\.14\.0
5\.14\.1


SKRIPT&#8203;:
\#\!/usr/bin/env perl

\*guard&#8203;::DESTROY = sub \{
&nbsp;   $\{$\_\[0\]\}\->\(\);
\};

\*guard = sub \(&\) \{
&nbsp;   my $callback = shift;
&nbsp;   return bless \\$callback\, "guard"
\};

my %t; %t = \(
&nbsp;    stash => \{
&nbsp;        guard => guard\(sub\{
&nbsp;            print "Guard\\n";
&nbsp;            delete $t\{stash\};
&nbsp;        \}\)\,
&nbsp;        foo => "bar"\,
&nbsp;        bar => "baz"\,
&nbsp;    \}\,
\);

delete $t\{stash\}\{guard\};

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.10.0:

Configured by Debian Project at Thu Jun 30 20:30:13 UTC 2011.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.32-5-686, archname=i486-linux-gnu-thread-multi
    uname='linux callisto 2.6.32-5-686 #1 smp mon jun 13 04:13:06 utc 2011 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.0 -Dsitearch=/usr/local/lib/perl/5.10.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=true, libperl=libperl.so.5.10.0
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /etc/perl
    /usr/local/lib/perl/5.10.0
    /usr/local/share/perl/5.10.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.0:
    HOME=/root
    LANG=de_DE.UTF-8
    LANGUAGE=de_DE:de:en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11
    PERL_BADLANG (unset)
    SHELL=/bin/bash

Flags:
    category=core
    severity=high

Site configuration information for perl 5.10.0:

Configured by Debian Project at Thu Jun 30 20:30:13 UTC 2011.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.32-5-686, archname=i486-linux-gnu-thread-multi
    uname='linux callisto 2.6.32-5-686 #1 smp mon jun 13 04:13:06 utc 2011 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.0 -Dsitearch=/usr/local/lib/perl/5.10.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=true, libperl=libperl.so.5.10.0
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /etc/perl
    /usr/local/lib/perl/5.10.0
    /usr/local/share/perl/5.10.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.0:
    HOME=/root
    LANG=de_DE.UTF-8
    LANGUAGE=de_DE:de:en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From felix.ostmann@thewar.de

i tested this snippet on 3 different servers/workstations \(debian
5\.0/debian 6\.0/ubuntu 10\.04\) with systemperl & selfbuilded perl

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

This appears to be related to \#99660\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Sep 29 00&#8203;:04&#8203;:29 2011\, ostmann@&#8203;websuche\.de wrote&#8203;:
> Attached pure\-perl script produce on all tested versions following
>    glibc error&#8203;:
> 
> \*\*\* glibc detected \*\*\* /home/websuche/perl5lib/bin/perl&#8203;:
>    munmap\_chunk\(\)&#8203;: invalid pointer&#8203;: 0x0000000000f72c48 \*\*\*
> 
> 
> TESTED VERSIONS&#8203;:
> 5\.8\.8
> 5\.10\.0
> 5\.10\.1
> 5\.14\.0
> 5\.14\.1
> 
> 
> SKRIPT&#8203;:
> \#\!/usr/bin/env perl
> 
> \*guard&#8203;::DESTROY = sub \{
>    $\{$\_\[0\]\}\->\(\);
> \};
> 
> \*guard = sub \(&\) \{
>    my $callback = shift;
>    return bless \\$callback\, "guard"
> \};
> 
> my %t; %t = \(
>     stash => \{
>         guard => guard\(sub\{
>             print "Guard\\n";
>             delete $t\{stash\};
>         \}\)\,
>         foo => "bar"\,
>         bar => "baz"\,
>     \}\,
> \);
> 
> delete $t\{stash\}\{guard\};

This has been fixed with commit 3b2cd80\.  If you need to work around it
on older perls\, call delete in non\-void context&#8203;:

&nbsp;    \(\) = delete $t\{stash\}\{guard\};

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
